pkg/controller/controller.go

```diff
@@ -197,56 +197,86 @@ func SetLatestApiVersionAnnotation(object metav1.Object) {
-				newDisk := request.AddVolumeOptions.Disk.DeepCopy()
-				newDisk.Name = request.AddVolumeOptions.Name
+func handleRemoveVolumeRequest(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	vmiSpec = removeVolume(vmiSpec, request)
+	vmiSpec = removeVolumeAsDisk(vmiSpec, request)
+	return vmiSpec
+}
 
-				vmiSpec.Domain.Devices.Disks = append(vmiSpec.Domain.Devices.Disks, *newDisk)
-			}
+func volumeAlreadyAdded(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) bool {
+	alreadyAdded := false
+	for _, volume := range vmiSpec.Volumes {
+		if volume.Name == request.AddVolumeOptions.Name {
+			alreadyAdded = true
+			break
 		}
+	}
 
-	} else if request.RemoveVolumeOptions != nil {
+	return alreadyAdded
+}
 
-		newVolumesList := []v1.Volume{}
-		newDisksList := []v1.Disk{}
+func addVolume(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	newVolume := &v1.Volume{
+		Name: request.AddVolumeOptions.Name,
+	}
 
-		for _, volume := range vmiSpec.Volumes {
-			if volume.Name != request.RemoveVolumeOptions.Name {
-				newVolumesList = append(newVolumesList, volume)
-			}
-		}
+	if request.AddVolumeOptions.VolumeSource.PersistentVolumeClaim != nil {
+		newVolume.VolumeSource.PersistentVolumeClaim = request.AddVolumeOptions.VolumeSource.PersistentVolumeClaim
+	} else if request.AddVolumeOptions.VolumeSource.DataVolume != nil {
+		newVolume.VolumeSource.DataVolume = request.AddVolumeOptions.VolumeSource.DataVolume
+	}
 
-		for _, disk := range vmiSpec.Domain.Devices.Disks {
-			if disk.Name != request.RemoveVolumeOptions.Name {
-				newDisksList = append(newDisksList, disk)
-			}
+	vmiSpec.Volumes = append(vmiSpec.Volumes, *newVolume)
+	return vmiSpec
+}
+
+func addVolumeAsDisk(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	if request.AddVolumeOptions.Disk != nil {
+		newDisk := request.AddVolumeOptions.Disk.DeepCopy()
+		newDisk.Name = request.AddVolumeOptions.Name
+
+		vmiSpec.Domain.Devices.Disks = append(vmiSpec.Domain.Devices.Disks, *newDisk)
+	}
+
+	return vmiSpec
+}
+
+func removeVolume(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	newVolumesList := []v1.Volume{}
+	for _, volume := range vmiSpec.Volumes {
+		if volume.Name != request.RemoveVolumeOptions.Name {
+			newVolumesList = append(newVolumesList, volume)
 		}
+	}
+	vmiSpec.Volumes = newVolumesList
+	return vmiSpec
+}
 
-		vmiSpec.Volumes = newVolumesList
-		vmiSpec.Domain.Devices.Disks = newDisksList
+func removeVolumeAsDisk(vmiSpec *v1.VirtualMachineInstanceSpec, request *v1.VirtualMachineVolumeRequest) *v1.VirtualMachineInstanceSpec {
+	newDisksList := []v1.Disk{}
+	for _, disk := range vmiSpec.Domain.Devices.Disks {
+		if disk.Name != request.RemoveVolumeOptions.Name {
```

Comment:
it:  In terms of performance, maybe it would be better to negate the condition - 
```
if disk.Name == request.RemoveVolumeOptions.Name {
    continue 
}
newDisksList = append(newDisksList, disk)
```

Reply:
Thanks Ezra.
Done.

Reply:
@ezrasilvera what performance you refer to?